File: TP1/haskell/fibonacci.hs
main :: IO ()
main = do
    print "fib at pos 10"
    print (fibonacci 10)

-- Exercise 8 | Calc fibonacci of a number

-- Naive.
fibonacciNaive :: Int -> Integer
fibonacciNaive 0 = 0
fibonacciNaive 1 = 1
fibonacciNaive x = fibonacciNaive(x-2) + fibonacciNaive(x-1)

-- Dynamic programming.
fibonacciSequence :: [Integer]
fibonacciSequence = go 0 1
    where go a b = a : go b (a+b)

fibonacci :: Int -> Integer
fibonacci x = fibonacciSequence !! x

File: TP1/prolog/fibonacci.pl
% Exercise 8 | Calc n-th fibonacci number [O(2^n)]

fibonacci(0, 0).
fibonacci(1, 1).
fibonacci(N, X) :-
    N > 0,
    N1 is N-1,
    N2 is N-2,
    fibonacci(N1, X1),
    fibonacci(N2, X2),
    X is X1+X2.
